docs/general: Add Sapphire Quickstart (#218)
oasisprotocol/docs: docs/general/developer-resources/sapphire-paratime/getting-started.md

# Getting Started on Sapphire

In this tutorial, you'll port an Eth project in under 10 minutes (or your money back),
and then go on to deploy a unique dapp that requires confidentiality to work.

> By the end of the tutorial, you should feel comfortable setting up your Eth development environment to target Sapphire, and know how and when to use confidentiality.

Reviewer: Trim to 80 characters per line. Also in the rest of the document.

The expected completion time of this tutorial is 30 minutes.

## Port an Eth project

### Setup

> Start by installing [truffle](https://github.com/trufflesuite/truffle#install).

Reviewer: ```suggestion
Start by installing [Truffle](https://github.com/trufflesuite/truffle#install).
```

> Then, run these commands in your terminal.

Reviewer: ```suggestion
Then, run these commands in your terminal:

```

> ```

Reviewer: ```suggestion
```sh
```

mkdir MetaCoin
truffle unbox MetaCoin
cd MetaCoin
git init
git add :/ && git commit -m "Initial commit"
pnpm init && pnpm add -D @truffle/hdwallet-provider # npm or yarn also works
```

### Deploy to Emerald (non-confidential EVM)

#### Get some eROSE

> 

Reviewer: ```suggestion

::: tip

To avoid confusion through the rest of this section we will denote ROSE the native token on the consensus layer only. When ROSE is deposited into Emerlad or Sapphire ParaTime we will call them eROSE or sROSE respectively.

:::

```

Set the second box to the address of a burner wallet.
It'll take a few moments to receive your eROSE after submitting the form.

#### Add the Emerald Testnet to Truffle

Apply this patch to `truffle-config.js`:

```diff
diff --git a/truffle-config.js b/truffle-config.js
index 68d534c..15c671d 100644
--- a/truffle-config.js
+++ b/truffle-config.js
@@ -22,7 +22,7 @@
 // const mnemonic = process.env["MNEMONIC"];
 // const infuraProjectId = process.env["INFURA_PROJECT_ID"];

-// const HDWalletProvider = require('@truffle/hdwallet-provider');
+const HDWalletProvider = require('@truffle/hdwallet-provider');

 module.exports = {
   /**
@@ -53,6 +53,14 @@ module.exports = {
     //   network_id: 5,       // Goerli's id
     //   chain_id: 5
     // }
+    emerald: {
+      provider: () =>
+        new HDWalletProvider([process.env.PRIVATE_KEY], "https://testnet.emerald.oasis.dev"),
+      network_id: 0xa515,
+    },
   },

   // Set default mocha options here, use special reporters etc.
```

#### Do the Truffle Thing

You'll want a script to run some methods on the contract.
Pop open your favorite editor and jam out the following stanza:
```javascript

> 

Reviewer: Is there something missing here?

```
Save it to `scripts/exercise-contract.js`.
We'll use it in just a bit.

Next, you can run the following and see the contract being deployed.

> ```

Reviewer: ```suggestion

```sh
```

PRIVATE_KEY="0x..." truffle migrate --network emerald
```
Everything should be smooth sailing so far.

Finally, run this line and observe some output.

> ```

Reviewer: ```suggestion

```
```

Hit up the one and only [Oasis Testnet faucet](https://faucet.testnet.oasis.dev/) and this time select "Sapphire".
Submit the form and your way.

#### Add the Sapphire Testnet to Truffle

And another diff for your applying pleasure:

```diff
diff --git a/truffle-config.js b/truffle-config.js
index 7af2f42..0cd9d36 100644
--- a/truffle-config.js
+++ b/truffle-config.js
@@ -58,6 +58,11 @@ module.exports = {
         new HDWalletProvider([process.env.PRIVATE_KEY], "https://testnet.emerald.oasis.dev"),
       network_id: 0xa515,
     },
+    sapphire: {
+      provider: () =>
+        new HDWalletProvider([process.env.PRIVATE_KEY], "https://testnet.sapphire.oasis.dev"),
+      network_id: 0x5aff,
+    },
   },

   // Set default mocha options here, use special reporters etc.
```

#### Port to Sapphire

Here's where things start to get interesting.
We're going to add confidentiality to this starter project in exactly two lines of code.

You'll need to grab the Sapphire compatibility library ([@oasisprotocol/sapphire-paratime](https://www.npmjs.com/package/@oasisprotocol/sapphire-paratime)), so make that happen by issuing a

> ```

Reviewer: ```suggestion

```sh
```

pnpm add @oasisprotocol/sapphire-paratime # npm also works
```

So far so good.
No import it by adding this line to the top of `truffle-config.js`:

> ```

Reviewer: ```suggestion

```js
```

const sapphire = require('@oasisprotocol/sapphire-paratime');
```
That's the first line of code.
Here's the second:

```diff
diff --git a/truffle-config.js b/truffle-config.js
index 0cd9d36..7db7cf8 100644
--- a/truffle-config.js
+++ b/truffle-config.js
@@ -60,7 +60,7 @@ module.exports = {
     },
     sapphire: {
       provider: () =>
-        new HDWalletProvider([process.env.PRIVATE_KEY], "https://testnet.sapphire.oasis.dev"),

> +        sapphire.wrap(new HDWalletProvider([process.env.PRIVATE_KEY], "https://testnet.sapphire.oasis.dev")),

Reviewer 2: Typescript says sapphire.wrap doesn't accept HDWalletProvider.
```sh
Argument of type 'HDWalletProvider' is not assignable to parameter of type 'UpstreamProvider'.
  Type 'HDWalletProvider' is not assignable to type '{ send: AsyncSend; }'.
    Types of property 'send' are incompatible.
      Type '(payload: JSONRPCRequestPayload, callback: (error: Error | null, response: JSONRPCResponsePayload) => void) => void' is not assignable to type 'AsyncSend'.
        Types of parameters 'payload' and 'args' are incompatible.
          Property 'jsonrpc' is missing in type 'Web3ReqArgs' but required in type 'JSONRPCRequestPayload'.
```

Author: what version of `sapphire-paratime` and `HDWalletProvider` are you using?

Author: Can you apply this diff and see if it works?
```diff
diff --git a/clients/js/src/compat.ts b/clients/js/src/compat.ts
index 3b49049..c357e8f 100644
--- a/clients/js/src/compat.ts
+++ b/clients/js/src/compat.ts
@@ -57,6 +57,7 @@ export type AsyncSend = (
 ) => void;

 export interface Web3ReqArgs {
+  readonly jsonrpc?: string;
   readonly id?: string | number;
   readonly method: string;
   readonly params?: any[];
 ```
 but why would TS be complaining about a field not existing on an object that doesn't request it? maybe that's not how `U extends Upstream` works and I need something closer to `deeply extends`

Reviewer 2: (that fix was attempted in https://github.com/oasisprotocol/sapphire-paratime/pull/46)

       network_id: 0x5aff,
     },
   },
```
This `wrap` function takes any kind of provider or signer you've got and turn it into one that works with Sapphire and confidentiality.
For the most part, wrapping your signer/provider is the least you'll need to do to get your dapp rollin' on Sapphire, but that's not a complete story since an unmodified contract may leak state through normal operation.

And now for the moment we've all been waiting for:

> ```

Reviewer: ```suggestion

```
```

Author: Thanks `matevz-lint`. I wish the machine did that for me so you don't have to :P

Reviewer: https://github.com/oasisprotocol/docs/issues/2 :wink:

Reviewer 2: Rules to enable for these:
https://github.com/oasisprotocol/docs/blob/f5d1b9cddf66dacc811a56e36bb31416baefcdee/.markdownlint-cli2.cjs#L12
https://github.com/oasisprotocol/docs/blob/f5d1b9cddf66dacc811a56e36bb31416baefcdee/.markdownlint-cli2.cjs#L22

(but even without those, `yarn lint-markdown` is currently unhappy)

### Run the Contract

And to wrap things up, we'll put `Vigil` through its paces.
First, let's see what's actually going on.

At the top of the file, we have our import du jour, `@oasisprotocol/sapphire-paratime`.
Unlike for Truffle, we have to "manually" wrap the signer since the Hardhat config only takes a private key. We do that at the top of `main`.

After deploying the contract, we can create a secret, check that it's not readable, wait a bit, and then check that it has become readable.
Pretty cool if you ask me!

Anyway, make it happen by running

```
PRIVATE_KEY="0x..." pnpm hardhat run scripts/run-vigil.ts --network sapphire
```

And if you see something like the following, you'll know you're well on the road to deploying confidential dapps on Sapphire.

```
Vigil deployed to: 0x74dC4879B152FDD1DDe834E9ba187b3e14f462f1
Storing a secret in 0x13125d868f5fb3cbc501466df26055ea063a90014b5ccc8dfd5164dc1dd67543
Checking the secret
failed to fetch secret: reverted: not expired
Waiting...
Checking the secret again
The secret ingredient is brussels sprouts
```

## All done!

Congratulations, you made it through the Sapphire tutorial.
Thanks for coming!

> If you have any more questions, please check out [the docs](https://docs.oasis.dev/general/developer-resources/sapphire-paratime) or the Discord community.

Reviewer: ```suggestion
If you have any more questions, please check out [the docs](README.mdx) or the Discord community.
```

> For more details on writing dapps for Sapphire, there are [docs](https://docs.oasis.dev/general/developer-resources/sapphire-paratime/writing-dapps-on-sapphire) for that, too.

Reviewer: ```suggestion
For more details on writing dapps for Sapphire, there are [docs](writing-dapps-on-sapphire.md) for that, too.
```
